feat(chart): Add resizePolicy support for ingress-nginx pods

younsl opened the pull request:
<!--- Provide a general summary of your changes in the Title above --->
<!--- Please don't @-mention people in PR or commit messages (do so in an additional comment). --->
<!--- Please make sure you title is descriptive, it is used in the Release notes to let others know what it does ---> 

## What this PR does / why we need it:

This PR adds support for Kubernetes resizePolicy configuration to the ingress-nginx Helm chart. The resizePolicy feature allows containers to be resized without pod restarts, enabling dynamic resource adjustments for better resource utilization and cost efficiency.

The implementation adds resizePolicy support for:
- Controller pods (Deployment and DaemonSet)
- Admission webhook jobs (createSecret and patchWebhook)

Related Kubernetes Documentation

- https://kubernetes.io/docs/concepts/workloads/pods/resize-pod-resources/
- https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#Container

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] CVE Report (Scanner found CVE and adding report)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

## Which issue/s this PR fixes
<!--
(optional, in `fixes #<issue number>` format, will close that issue when PR gets merged):

fixes #
-->

## How Has This Been Tested?

  - Added comprehensive unit tests for all resizePolicy configurations
  - All tests pass successfully:

```bash
$ helm unittest .
### Chart [ ingress-nginx ] .

 PASS  Controller > ConfigMap > Add Headers     tests/controller-configmap-addheaders_test.yaml
 PASS  Controller > ConfigMap > Proxy Headers   tests/controller-configmap-proxyheaders_test.yaml
 PASS  Controller > ConfigMap   tests/controller-configmap_test.yaml
 PASS  Controller > DaemonSet   tests/controller-daemonset_test.yaml
 PASS  Controller > Deployment  tests/controller-deployment_test.yaml
 PASS  Controller > HPA tests/controller-hpa_test.yaml
 PASS  Controller > IngressClass > Aliases      tests/controller-ingressclass-aliases_test.yaml
 PASS  Controller > IngressClass        tests/controller-ingressclass_test.yaml
 PASS  Controller > KEDA        tests/controller-keda_test.yaml
 PASS  Controller > NetworkPolicy       tests/controller-networkpolicy_test.yaml
 PASS  Controller > PodDisruptionBudget tests/controller-poddisruptionbudget_test.yaml
 PASS  Controller > PrometheusRule      tests/controller-prometheusrule_test.yaml
 PASS  Controller > Service > Internal  tests/controller-service-internal_test.yaml
 PASS  Controller > Service > Metrics   tests/controller-service-metrics_test.yaml
 PASS  Controller > Service > Webhook   tests/controller-service-webhook_test.yaml
 PASS  Controller > Service     tests/controller-service_test.yaml
 PASS  Controller > ServiceAccount      tests/controller-serviceaccount_test.yaml
 PASS  Controller > ServiceMonitor      tests/controller-servicemonitor_test.yaml
 PASS  Default Backend > Deployment     tests/default-backend-deployment_test.yaml
 PASS  Default Backend > Extra ConfigMaps       tests/default-backend-extra-configmaps_test.yaml
 PASS  Default Backend > PodDisruptionBudget    tests/default-backend-poddisruptionbudget_test.yaml
 PASS  Default Backend > Service        tests/default-backend-service_test.yaml
 PASS  Default Backend > ServiceAccount tests/default-backend-serviceaccount_test.yaml

Charts:      1 passed, 1 total
Test Suites: 23 passed, 23 total
Tests:       138 passed, 138 total
Snapshot:    0 passed, 0 total
Time:        1.075678208s
```
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [x] I have added unit and/or e2e tests to cover my changes.
- [x] All new and existing tests passed.

younsl commented:
This PR is similar to https://github.com/oauth2-proxy/manifests/pull/346

Gacko reviewed (changes requested):
Since this feature is supported from v1.30 upwards only, I'd ask you to wrap it into a condition to check if the target cluster supports it or not.

Gacko reviewed (changes requested):
Please do not re-request my review without having changed anything. Thank you! 😄

younsl commented:
@Gacko Thanks for the feedback! I've updated the implementation to include Kubernetes version checks for the `resizePolicy` feature.

Gacko commented on charts/ingress-nginx/templates/controller-daemonset.yaml:
```diff
@@ -174,6 +174,9 @@ spec:
         {{- if .Values.controller.resources }}
           resources: {{ toYaml .Values.controller.resources | nindent 12 }}
         {{- end }}
+        {{- if and .Values.controller.resizePolicy (semverCompare ">=1.30.0-0" .Capabilities.KubeVersion.Version) }}
```
Please have a look how we are implementing such capability checks elsewhere in the chart and align yours to these.

Gacko commented on charts/ingress-nginx/templates/controller-deployment.yaml:
```diff
@@ -180,6 +180,9 @@ spec:
         {{- if .Values.controller.resources }}
           resources: {{ toYaml .Values.controller.resources | nindent 12 }}
         {{- end }}
+        {{- if and .Values.controller.resizePolicy (semverCompare ">=1.30.0-0" .Capabilities.KubeVersion.Version) }}
```
Please have a look how we are implementing such capability checks elsewhere in the chart and align yours to these.

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/job-patch/job-createSecret_test.yaml:
```diff
@@ -76,3 +76,46 @@ tests:
                             apiVersion: v1
                             fieldPath: metadata.namespace
+
+  - it: should create a Job with resizePolicy if `controller.admissionWebhooks.createSecretJob.resizePolicy` is set and Kubernetes version >= 1.30
```
```suggestion
  - it: should create a Job with resize policy if `controller.admissionWebhooks.createSecretJob.resizePolicy` is set
```

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/job-patch/job-createSecret_test.yaml:
```diff
@@ -76,3 +76,46 @@ tests:
+          path: spec.template.spec.containers[0].resizePolicy[1].restartPolicy
+          value: RestartContainer
+
+  - it: should NOT create a Job with resizePolicy if `controller.admissionWebhooks.createSecretJob.resizePolicy` is set but Kubernetes version < 1.30
+    capabilities:
+      majorVersion: 1
+      minorVersion: 29
+    set:
+      controller.admissionWebhooks.createSecretJob.resizePolicy:
+        - resourceName: cpu
+          restartPolicy: NotRequired
+    asserts:
+      - notExists:
+          path: spec.template.spec.containers[0].resizePolicy
+
+  - it: should not include resizePolicy if `controller.admissionWebhooks.createSecretJob.resizePolicy` is empty
+    set:
+      controller.admissionWebhooks.createSecretJob.resizePolicy: []
+    asserts:
+      - notExists:
+          path: spec.template.spec.containers[0].resizePolicy
```
I don't think there is a need for that.

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/job-patch/job-patchWebhook_test.yaml:
```diff
@@ -76,3 +76,46 @@ tests:
                             apiVersion: v1
                             fieldPath: metadata.namespace
+
+  - it: should create a Job with resizePolicy if `controller.admissionWebhooks.patchWebhookJob.resizePolicy` is set and Kubernetes version >= 1.30
```
```suggestion
  - it: should create a Job with resize policy if `controller.admissionWebhooks.patchWebhookJob.resizePolicy` is set
```

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/job-patch/job-patchWebhook_test.yaml:
```diff
@@ -76,3 +76,46 @@ tests:
                             fieldPath: metadata.namespace
+
+  - it: should create a Job with resizePolicy if `controller.admissionWebhooks.patchWebhookJob.resizePolicy` is set and Kubernetes version >= 1.30
+    capabilities:
+      majorVersion: 1
+      minorVersion: 30
```
Have you checked if this is required? I am actually not 100% sure, I can remember we needed that in some other test, but I can barely remember without checking.

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/job-patch/job-patchWebhook_test.yaml:
```diff
@@ -76,3 +76,46 @@ tests:
+          path: spec.template.spec.containers[0].resizePolicy[1].restartPolicy
+          value: RestartContainer
+
+  - it: should NOT create a Job with resizePolicy if `controller.admissionWebhooks.patchWebhookJob.resizePolicy` is set but Kubernetes version < 1.30
+    capabilities:
+      majorVersion: 1
+      minorVersion: 29
+    set:
+      controller.admissionWebhooks.patchWebhookJob.resizePolicy:
+        - resourceName: cpu
+          restartPolicy: NotRequired
+    asserts:
+      - notExists:
+          path: spec.template.spec.containers[0].resizePolicy
+
+  - it: should not include resizePolicy if `controller.admissionWebhooks.patchWebhookJob.resizePolicy` is empty
+    set:
+      controller.admissionWebhooks.patchWebhookJob.resizePolicy: []
+    asserts:
+      - notExists:
+          path: spec.template.spec.containers[0].resizePolicy
```
I don't think there is a need for that.

Gacko commented on charts/ingress-nginx/tests/controller-daemonset_test.yaml:
```diff
@@ -208,3 +208,49 @@ tests:
           path: spec.template.spec.runtimeClassName
           value: myClass
+
+  - it: should create a DaemonSet with resizePolicy if `controller.resizePolicy` is set and Kubernetes version >= 1.30
```
```suggestion
  - it: should create a DaemonSet with resize policy if `controller.resizePolicy` is set
```

Gacko commented on charts/ingress-nginx/tests/controller-daemonset_test.yaml:
```diff
@@ -208,3 +208,49 @@ tests:
+        - resourceName: memory
+          restartPolicy: RestartContainer
+    asserts:
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[0].resourceName
+          value: cpu
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[0].restartPolicy
+          value: NotRequired
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[1].resourceName
+          value: memory
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[1].restartPolicy
+          value: RestartContainer
```
`value` can also be a whole array. You do not need to check each element of the array.

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/job-patch/job-patchWebhook_test.yaml:
```diff
@@ -76,3 +76,46 @@ tests:
+        - resourceName: memory
+          restartPolicy: RestartContainer
+    asserts:
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[0].resourceName
+          value: cpu
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[0].restartPolicy
+          value: NotRequired
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[1].resourceName
+          value: memory
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[1].restartPolicy
+          value: RestartContainer
```
`value` can also be a whole array. You do not need to check each element of the array.

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/job-patch/job-createSecret_test.yaml:
```diff
@@ -76,3 +76,46 @@ tests:
+        - resourceName: memory
+          restartPolicy: RestartContainer
+    asserts:
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[0].resourceName
+          value: cpu
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[0].restartPolicy
+          value: NotRequired
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[1].resourceName
+          value: memory
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[1].restartPolicy
+          value: RestartContainer
```
`value` can also be a whole array. You do not need to check each element of the array.

Gacko commented on charts/ingress-nginx/tests/controller-daemonset_test.yaml:
```diff
@@ -208,3 +208,49 @@ tests:
+          path: spec.template.spec.containers[0].resizePolicy[1].restartPolicy
+          value: RestartContainer
+
+  - it: should NOT create a DaemonSet with resizePolicy if `controller.resizePolicy` is set but Kubernetes version < 1.30
+    capabilities:
+      majorVersion: 1
+      minorVersion: 29
+    set:
+      controller.kind: DaemonSet
+      controller.resizePolicy:
+        - resourceName: cpu
+          restartPolicy: NotRequired
+    asserts:
+      - notExists:
+          path: spec.template.spec.containers[0].resizePolicy
+
+  - it: should not include resizePolicy if `controller.resizePolicy` is empty
+    set:
+      controller.kind: DaemonSet
+      controller.resizePolicy: []
+    asserts:
+      - notExists:
+          path: spec.template.spec.containers[0].resizePolicy
```
I don't think there is a need for that.

Gacko commented on charts/ingress-nginx/tests/controller-deployment_test.yaml:
```diff
@@ -231,3 +231,46 @@ tests:
           path: spec.template.spec.runtimeClassName
           value: myClass
+
+  - it: should create a Deployment with resizePolicy if `controller.resizePolicy` is set and Kubernetes version >= 1.30
```
```suggestion
  - it: should create a Deployment with resize policy if `controller.resizePolicy` is set
```

Gacko commented on charts/ingress-nginx/tests/controller-daemonset_test.yaml:
```diff
@@ -208,3 +208,49 @@ tests:
           value: myClass
+
+  - it: should create a DaemonSet with resizePolicy if `controller.resizePolicy` is set and Kubernetes version >= 1.30
+    capabilities:
+      majorVersion: 1
+      minorVersion: 30
```
Have you checked if this is required? I am actually not 100% sure, I can remember we needed that in some other test, but I can barely remember without checking.

Gacko commented on charts/ingress-nginx/tests/controller-deployment_test.yaml:
```diff
@@ -231,3 +231,46 @@ tests:
+        - resourceName: memory
+          restartPolicy: RestartContainer
+    asserts:
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[0].resourceName
+          value: cpu
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[0].restartPolicy
+          value: NotRequired
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[1].resourceName
+          value: memory
+      - equal:
+          path: spec.template.spec.containers[0].resizePolicy[1].restartPolicy
+          value: RestartContainer
```
`value` can also be a whole array. You do not need to check each element of the array.

younsl commented:
@Gacko PTAL

Gacko commented on charts/ingress-nginx/values.yaml:
```diff
@@ -401,6 +401,13 @@ controller:
+  # -- Container resizePolicy for CPU and memory
+  # ref: https://kubernetes.io/docs/concepts/workloads/pods/user-namespaces/
+  resizePolicy: []
+    # - resourceName: cpu
+    #   restartPolicy: NotRequired
+    # - resourceName: memory
+    #   restartPolicy: RestartContainer
```
```suggestion
  # - resourceName: cpu
  #   restartPolicy: NotRequired
  # - resourceName: memory
  #   restartPolicy: RestartContainer
```

Gacko commented on charts/ingress-nginx/values.yaml:
```diff
@@ -795,6 +802,12 @@ controller:
       # requests:
       #   cpu: 10m
       #   memory: 20Mi
+      # -- Container resizePolicy for createSecret job
```
```suggestion
      # -- Resize policy for secret creation containers
```

Gacko replied:
And please also add docs here, if you add them for the controller.

Gacko commented on charts/ingress-nginx/values.yaml:
```diff
@@ -795,6 +802,12 @@ controller:
       #   memory: 20Mi
+      # -- Container resizePolicy for createSecret job
+      resizePolicy: []
+        # - resourceName: cpu
+        #   restartPolicy: NotRequired
+        # - resourceName: memory
+        #   restartPolicy: RestartContainer
```
```suggestion
      # - resourceName: cpu
      #   restartPolicy: NotRequired
      # - resourceName: memory
      #   restartPolicy: RestartContainer
```

Gacko commented on charts/ingress-nginx/values.yaml:
```diff
@@ -822,6 +835,12 @@ controller:
             - ALL
         readOnlyRootFilesystem: true
       resources: {}
+      # -- Container resizePolicy for patchWebhook job
```
```suggestion
      # -- Resize policy for webhook patch containers
```

Gacko replied:
And please also add docs here, if you add them for the controller.

Gacko commented on charts/ingress-nginx/values.yaml:
```diff
@@ -822,6 +835,12 @@ controller:
       resources: {}
+      # -- Container resizePolicy for patchWebhook job
+      resizePolicy: []
+        # - resourceName: cpu
+        #   restartPolicy: NotRequired
+        # - resourceName: memory
+        #   restartPolicy: RestartContainer
```
```suggestion
      # - resourceName: cpu
      #   restartPolicy: NotRequired
      # - resourceName: memory
      #   restartPolicy: RestartContainer
```

younsl commented:
@Gacko Thank you for the thorough review. I've addressed all your feedback:

- Updated capability check to v1.33 for beta feature
- Removed resizePolicy from admission webhook jobs (short-lived, <5s)
- Aligned capability checks with existing chart patterns
- Fixed indentation and updated documentation

Gacko reviewed (approved):
/triage accepted
/kind feature
/priority backlog
/lgtm

I rebased your branch on top of `main`, just in case you wonder about the force push. Also I squashed your commits. Thank you very much for your endurance and efforts! 🙂

Gacko commented:
/unhold
